crates/pipeline_manager/src/pipeline_manager.rs

```diff
@@ -881,78 +859,45 @@ struct CompileProgramRequest {
             , body = ErrorResponse
             , example = json!(example_outdated_program_version())),
     ),
-    tag = "Program"
-)]
-#[post("/programs/compile")]
-async fn compile_program(
-    state: WebData<ServerState>,
-    tenant_id: ReqData<TenantId>,
-    request: web::Json<CompileProgramRequest>,
-) -> Result<HttpResponse, DBError> {
-    state
-        .db
-        .lock()
-        .await
-        .set_program_pending(*tenant_id, request.program_id, request.version)
-        .await?;
-
-    Ok(HttpResponse::Accepted().finish())
-}
-
-/// Request to cancel ongoing program compilation.
-#[derive(Deserialize, ToSchema)]
-struct CancelProgramRequest {
-    /// Program id.
-    program_id: ProgramId,
-    /// Latest program version known to the client.
-    version: Version,
-}
-
-/// Cancel outstanding compilation request.
-///
-/// The client should poll the `/program_status` endpoint
-/// to determine when the cancelation request completes.
-#[utoipa::path(
-    request_body = CancelProgramRequest,
-    responses(
-        (status = ACCEPTED, description = "Cancelation request submitted."),
-        (status = NOT_FOUND
-            , description = "Specified program id does not exist in the database."
-            , body = ErrorResponse
-            , example = json!(example_unknown_program())),
-        (status = CONFLICT
-            , description = "Program version specified in the request doesn't match the latest program version in the database."
-            , body = ErrorResponse
-            , example = json!(example_outdated_program_version())),
+    params(
+        ("program_id" = Uuid, Path, description = "Unique program identifier")
     ),
     tag = "Program"
 )]
-#[delete("/programs/compile")]
-async fn cancel_program(
+#[post("/programs/{program_id}/compile")]
+async fn compile_program(
     state: WebData<ServerState>,
     tenant_id: ReqData<TenantId>,
-    request: web::Json<CancelProgramRequest>,
-) -> Result<HttpResponse, DBError> {
+    request: HttpRequest,
+    body: web::Json<CompileProgramRequest>,
+) -> Result<HttpResponse, ManagerError> {
+    let program_id = ProgramId(parse_uuid_param(&request, "program_id")?);
     state
         .db
         .lock()
         .await
-        .cancel_program(*tenant_id, request.program_id, request.version)
+        .set_program_pending(*tenant_id, program_id, body.version)
         .await?;
 
     Ok(HttpResponse::Accepted().finish())
 }
 
 /// Delete a program.
```

Comment:
It's probably a good idea to document the behavior in more detail (fails if there exists at least one pipeline associated with the program).

Reply:
Oh, I didn't know about that behavior. So if there is a pipeline running a program, and in the meantime we edit the program, we can't compile the new version of the program (to see if it's safe to commit to a deployment etc)?